### PR TITLE
Tests for sparse implementation of topographica

### PR DESCRIPTION
**Pull request by Tasignotas**
Implementation of #596 

I have made some progress on writing the tests. However, the test fails, because the output after the first iteration doesn't match with the expected output. Is there something obviously wrong with my code?

### CODE REVIEW

File: topo/sparse/sparse.pyx [hunk 1]
```diff
@@ -260,6 +260,16 @@ cdef class csarray_float:
...
 
+    def toSparseArray(self):
+        """
+        Convert this sparse matrix into scipy.sparse CSR matrix.
+        """
+        cdef numpy.ndarray[int, ndim=1, mode="c"] rowInds
+        cdef numpy.ndarray[int, ndim=1, mode="c"] colInds
+        cdef numpy.ndarray[float, ndim=1, mode="c"] values
+        rowInds, colInds, values = self.getTriplets()
+        return scipy.sparse.csr_matrix((values, (rowInds, colInds)), shape=self.shape)
```
Comment (jbednar) [inline topo/sparse/sparse.pyx]: Is this the only dependence on SciPy?  Is it possible to use topo.sparse without Scipy?  if it's possible, please move the import of scipy into this function.

**Comment by Tasignotas** [comment]
As suggested by @jbednar, I moved the construction of `SparseGCAL` test model into `gcal_sparse.ty` and also improved the decorator registering procedure which works fine for all training tests.

The `SparseGCAL` currently only switches out the `V1_afferent` to use `SparseCFProjection`. The results match the ones produced by the dense GCAL (gcal_sparse.ty_DATA is a copy of gcal.ty_DATA).

**Comment by Tasignotas** [comment]
Had to rebase the branch. Seems like it should be possible to pull my changes now.

**Comment by Tasignotas** [comment]
Had to rebase the branch. Seems like it should be possible to pull my changes now.

**Comment by jlstevens** [comment]
Looks great! This PR now addresses just the tests as it should. Personally, I would like to see the GPU PR now, merge that into master then merge these tests.

**Comment by philippjfr** [comment]
Agreed, this looks good. I think this branch should be merged first though if merely so you don't have to do some complex rebasing to get them in the right order.

**Comment by jlstevens** [comment]
Ah right. I keep forgetting this isn't GPU tests but sparse tests (which Philipp and I should have added a long time ago!). So yes, this should be merged and thanks again for doing this!

**Comment by jbednar** [comment]
The final result looks great, and I am happy for those changes to be made to Topographica.  

Philipp, is there any way to cleanly apply the final changes only, so that we don't get the complete history of these changes in Topographica's history?  The net result is a very small set of self-contained changes, clearly not disruptive of the rest of the codebase, but the sum of all the individual changes is many times larger and more confusing.  So ideally, we'd just get the final set of changes, which would be fine as a single commit.  However, if that would cause problems with Ignotas' GPU code, which was based on this branch, we can just bite the bullet and merge this branch as-is.  Can you please make that judgement call and either hit Merge on this PR or else create a commit with the net changes and push that?  Thanks.

**Comment by philippjfr** [comment]
I don't want to steal Ignotas attribution for these commits. What I could do is clone his branch, perform an interactive rebase to collapse it into a single commit and then push that but I don't think it's worth the effort and the headaches it'll cause Ignotas. No one is going to go through these commits one by one and look at the changes and in the grand scheme of things 12 commits is nothing. So I'll just go ahead and merge this now.

**Comment by Tasignotas** [comment]
Thanks!

**Comment by jbednar** [comment]
That's what I suspected.  Thanks for checking and for merging!

**Pull request merged**